tokio/src/task/task_local.rs

```diff
@@ -332,6 +332,76 @@ pin_project! {
     }
 }
 
+impl<T, F> TaskLocalFuture<T, F>
+where
+    T: 'static,
+{
+    /// Takes the task local value `T` owned by the `TaskLocalFuture`. If the
+    /// task local value exists, then returns `Some(T)` and the task local value
+    /// inside the `TaskLocalFuture` becomes unset. If it does not exist,
+    /// it returns `None`.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # async fn dox() {
+    ///  tokio::task_local! {
+    ///     static KEY: u32;
+    ///  }
+    ///
+    ///  let fut = KEY.scope(42, async {
+    ///     // Do some async work
+    ///  });
+    ///
+    ///  let mut pinned = Box::pin(fut);
+    ///
+    ///  // Complete the TaskLocalFuture
+    ///  let _ = (&mut pinned).as_mut().await;
+    ///
+    ///  // And here, we can take task local value
+    ///  let value = pinned.as_mut().take_value();
+    ///
+    ///  assert_eq!(value, Some(42));
+    /// # }
+    /// ```
+    ///
+    /// # Note
+    ///
+    /// Note that this function attempts to take the task local value regardless of
+    /// whether the `TaskLocalFuture` is completed or not. This means that if you
+    /// call this function before the `TaskLocalFuture` is completed, you need to
+    /// make sure that the access to the task local value in the `TaskLocalFuture` is safe.
+    ///
+    /// For example, the following code returns `Err` for accessing the `KEY` variable
+    /// in the async block of the `scope` function.
+    ///
+    /// ```
+    /// # async fn dox() {
+    /// tokio::task_local! {
+    ///    static KEY: u32;
+    /// }
+    ///
+    /// let fut = KEY.scope(42, async {
+    ///    // Since `take_value()` has already been called at this point,
+    ///    // `try_with` here will fail.
+    ///    assert!(KEY.try_with(|_| {}).is_err())
+    /// });
+    ///
+    /// let mut pinned = Box::pin(fut);
+    ///
+    /// // With this call, the task local value of fut is unset.
+    /// assert_eq!(pinned.as_mut().take_value(), Some(42));
+    ///
+    /// // Poll **after** invoking `take_value()`
+    /// let _ = (&mut pinned).as_mut().await;
+    /// # }
+    /// ```
+    pub fn take_value(self: Pin<&mut Self>) -> Option<T> {
```

Comment:
The change itself looks good, but I'd like the documentation to be a bit improved.

A few thoughts:

* The first line of documentation should be a short one-line summary of what it does. The body of the documentation then follows after an empty line. This could be e.g. "Returns the value stored in the task local by this future."
* We generally don't have sections after the examples, so your "Note" section should probably be moved. It can be combined with the body you already have at the top.
* For a function like this, a bulleted list often works well for explaining the behavior in each case.
* I would probably put the note about what happens if you poll the future after calling this method after the bulleted list.
* Your examples have one space too many. Currently, they will appear in the documentation indented by one space.

Reply:
Updated. 
The code snippets in the note section might be redundant? I couldn't find documentation that use code snippets other than example.

Reply:
Yeah, we probably only need one example.